### ISSUE TITLE
NoClassDefFoundError org.eclipse.jetty.io.SelectorManager seen on Android sentry crashlytics backend

### ISSUE THREAD

**Issue by courville** [jupnp/jupnp#236]
When deploying jupnp on my android application nova video player (https://github.com/nova-video-player/aos-AVP), I see on my sentry crashlytics backend the following error for some users (I do not experience it myself):
```
java.lang.NoClassDefFoundError: org.eclipse.jetty.io.SelectorManager$$ExternalSyntheticLambda1
    at org.eclipse.jetty.io.SelectorManager.<init>(SourceFile:100)
    at org.eclipse.jetty.server.ServerConnector$ServerConnectorManager.<init>(SourceFile:532)
    at org.eclipse.jetty.server.ServerConnector.newSelectorManager(SourceFile:227)
    at org.eclipse.jetty.server.ServerConnector.<init>(SourceFile:220)
    at org.eclipse.jetty.server.ServerConnector.<init>(SourceFile:94)
    at org.jupnp.transport.impl.jetty.JettyServletContainer.addConnector(SourceFile:65)
    at org.jupnp.transport.impl.ServletStreamServerImpl.init(SourceFile:61)
    at org.jupnp.transport.RouterImpl.startAddressBasedTransports(SourceFile:403)
    at org.jupnp.transport.RouterImpl.enable(SourceFile:127)
    at org.jupnp.android.AndroidRouter.enable(SourceFile:88)
    at org.jupnp.UpnpServiceImpl.startup(SourceFile:263)
    at com.archos.mediacenter.filecoreextension.upnp2.UpnpServiceManager$2.onServiceConnected(SourceFile:274)
    at android.app.LoadedApk$ServiceDispatcher.doConnected(LoadedApk.java:1244)
    at android.app.LoadedApk$ServiceDispatcher$RunConnection.run(LoadedApk.java:1261)
    at android.os.Handler.handleCallback(Handler.java:739)
    at android.os.Handler.dispatchMessage(Handler.java:95)
    at android.os.Looper.loop(Looper.java:148)
    at android.app.ActivityThread.main(ActivityThread.java:5431)
    at java.lang.reflect.Method.invoke(Method.java)
    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:810)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:698)
```
Perhaps this is due to using an older version of jetty in order to not target Android API26 since my application is targeting API21 (https://github.com/nova-video-player/aos-MediaLib/blob/v6.2/build.gradle):
```
    api 'org.jupnp:org.jupnp:3.0.2'
    api 'org.jupnp:org.jupnp.support:3.0.2'
    api 'org.jupnp:org.jupnp.android:3.0.2'
    // jetty required by jupnp and going above 9.4.14 requires minSdkVersion 26
    implementation ('org.eclipse.jetty:jetty-server:9.4.14.v20181114') {
        exclude group: 'org.eclipse.jetty.orbit', module: 'javax.servlet'
    }
    implementation ('org.eclipse.jetty:jetty-servlet:9.4.14.v20181114') {
        exclude group: 'org.eclipse.jetty.orbit', module: 'javax.servlet'
    }
    implementation ('org.eclipse.jetty:jetty-client:9.4.14.v20181114') {
        exclude group: 'org.eclipse.jetty.orbit', module: 'javax.servlet'
    }
```
Please let me know if there is something obvious that I miss.

**Comment by wborn** [comment]
I see you have a similar issue with https://github.com/hierynomus/smbj/issues/807.

Apparently Java generated a class for this lambda:

https://github.com/jetty/jetty.project/blob/c4550056e785fb5665914545889f21dc136ad9e6/jetty-io/src/main/java/org/eclipse/jetty/io/SelectorManager.java#L100

Maybe you are using ProGuard and it is configured to optimize these classes away to reduce the code?

**Comment by courville** [comment]
Indeed I forgot about this one (smbj). Thanks: I will work on the proguard config.

**Comment by wborn** [comment]
I asked an AI what ExternalSyntheticLambda classes are and this is what it came up with:

---

ExternalSyntheticLambda classes are classes that Android Studio generates automatically when you use lambda expressions in your Android code. 

Here's a breakdown of the terms:

* **External:** These classes are separate from your regular activity or fragment classes.
* **Synthetic:** They are auto-generated by Android Studio, so you won't find them in your source code.
* **Lambda:** Lambdas are a concise way to define anonymous methods in Java and Kotlin. They are often used for short, single-expression functions like event listeners.

When you use a lambda expression in your Android code, Android Studio translates it behind the scenes into a call to a method in an ExternalSyntheticLambda class. This class contains the actual implementation of the logic you defined in the lambda.

Here are some key points to remember about ExternalSyntheticLambda classes:

* You don't need to directly interact with these classes. They are for internal use by Android Studio.
* These classes might be renamed or removed in future versions of Android Studio, so it's best to avoid relying on their specific implementation details.
* If you're using a decompiler tool to look at your app's code, you might see these ExternalSyntheticLambda classes. They can be helpful for understanding how your lambdas are being translated, but it's generally not necessary to modify them.

In summary, ExternalSyntheticLambda classes are a behind-the-scenes detail of how Android Studio handles lambda expressions. You can focus on writing concise and expressive code with lambdas, and Android Studio takes care of the rest.

**Comment by wborn** [comment]
I don't see these classes in the Java JAR (https://repo1.maven.org/maven2/org/eclipse/jetty/jetty-io/9.4.11.v20180605/jetty-io-9.4.11.v20180605.jar), so they are probably generated by the Android toolchain. Maybe updating the toolchain helps with fixing bugs.

**Comment by wborn** [comment]
https://github.com/aws-amplify/amplify-flutter/issues/1239 also looks similar.

**Comment by courville** [comment]
Thank @wborn for the help and links.

**Comment by courville** [comment]
FTR I reproduce systematically the issue with an old Android emulator image (Android 5.0 API21) but no recent Android versions (e.g. 14, 13, 12, 11).
However the issue is gone when using:
```
compileOptions {
        coreLibraryDesugaringEnabled true
        sourceCompatibility JavaVersion.VERSION_1_8
        targetCompatibility JavaVersion.VERSION_1_8
    }
dependencies {
    coreLibraryDesugaring 'com.android.tools:desugar_jdk_libs:2.0.4'
```
instead of `JavaVersion.VERSION_17`.

**Comment by courville** [comment]
Despite what the documentation claims https://developer.android.com/build/jdks#groovy it seems that `JavaVersion.VERSION_17` can be used with `coreLibraryDesugaringEnabled true` and it avoids the issues with `ExternalSyntheticLambda` (at least on my experiments).
